README.md
<a href="https://shortlist-digital.github.io/tapestry-wp"><img src="http://i.imgur.com/HtzivRT.png" height="120" /></a>

## WordPress theming with React ⚛

* Unbelievably fast. 
* Client side SPA routing with React Router.
* Glamor CSS.
* Code splitting out of the box.
* Server-side-rendered.
* Minimal configuration.
* WordPress for your CMS.
---

[![Greenkeeper badge](https://badges.greenkeeper.io/shortlist-digital/tapestry-wp.svg)](https://greenkeeper.io/)
[![npm](https://img.shields.io/npm/v/tapestry-wp.svg)]()
[![CircleCI](https://circleci.com/gh/shortlist-digital/tapestry-wp/tree/develop.svg?style=shield)](https://circleci.com/gh/shortlist-digital/tapestry-wp/tree/develop)
[![Known Vulnerabilities](https://snyk.io/test/github/shortlist-digital/tapestry-wp/badge.svg)](https://snyk.io/test/github/shortlist-digital/tapestry-wp)

Tapestry is a Universal JavaScript Application for rendering React front-ends via the Wordpress API.

## Quick Start

Install  `tapestry-wp` through `npm` or `yarn`
```text
npm install tapestry-wp --save
```
Map the Tapestry commands to `npm` scripts in your `package.json`
```js
"scripts": {
  "start": "tapestry",
  "start:prod": "tapestry start",
  "bootstrap": "tapestry init"
}
```
You can now either run `npm run bootstrap` to create a simple Tapestry project or manually add a `tapestry.config.js` to your project root.
```js
import Post from './components/post'
import Page from './components/page'

export default {
  components: {
    Post, Page
  },
  siteUrl: 'http://your-wordpress.url'
}
```

### API
* `siteUrl`: The Wordpress instance to access the WP-API. e.g. `http://thesun.co.uk`
* `components`: An object with React components mapped to Wordpress endpoints. e.g. `Post`, `Page`, `Term`, `Category`
* `loaders`: An optional object with data loading functions matching components.
* `proxyPaths`: An array of paths to allow proxy access e.g. `['robots.txt', 'favicon.ico']`
* `host`: The host Tapestry is assigned to. e.g. `localhost`
* `port`: The port Tapestry is assigned to. e.g. `3030`
* `onPageUpdate`: A function called on React-router `route` event

### Plugins
If running WordPress 4.7 or later, the [Rest Filter WordPress plugin](https://github.com/WP-API/rest-filter) will need to be installed to run `tapestry-wp`.

## Roadmap
Tapestry has a long list of features that we are looking to implement, including confirmed and super speculative features.
- [x] Server render React component tree
- [x] Bundle the app for the client
- [x] Provide production option for client bundle
- [x] Supply CLI to run the server
- [x] Hook up WP-API to match routes
- [x] Render `<head>` tags with `react-helmet`
- [x] Supply default routes
- [x] Integrate Glamor CSS-in-JS framework, rendering server-side, hydrating on the client
- [x] Ability to override data loaders
- [ ] Ability to override routing
- [ ] WordPress plugin to enable previewing, 'View Post' functionality and permalinks
- [ ] Create a global Redux store of page, post and archive data
- [ ] Handle Redirects
- [ ] Access options, menus through WP-API


CHANGELOG.md
#### 4.0.0

* Custom Document Support
* Multiple endpoint purging
* Polyfills for React 16
* Redirects handler runs at the beginning of a request
* File loader in webpack for hashing off static assets
* Housekeeping of files

#### 3.0.0

* Release 🎉

#### 3.0.0-10
#### 3.0.0-9
#### 3.0.0-8

* Remove over-zealous logging

#### 3.0.0-8

* Remove giant asyncProps log

#### 3.0.0-7

* Refactor proxy handlers to prevent timeout errors on non 2xx requests
* Skip cache if preview (even though there shouldn't be anything in the cache)

#### 3.0.0-6

* Increase keepalive timeout, extend request timeout

#### 3.0.0-5

* Add fetcher shared file to allow for connection pooling on the server

#### 3.0.0-1
* Configurable Redis Cache 

#### 3.0.0-0
* Cluster management

#### 2.3.1
* Redirects are case insensitive

#### 2.3.0
* Add CACHE_MAX_ITEM_COUNT var to increase cache size

#### 2.2.0
* Add cachebust param to redirects json request

#### 2.1.0
* Add `tapestry hot` command for live reloading of client

#### 2.0.0
* Homegenise how `ENV` variables are used
* Add `wordpress.com` site support

#### 2.0.0-26
* Rewrite of how redirects are handled - improves boot time with a large number of redirects

#### 2.0.0-25
* Fixed missing rendering of Hapi redirect errors

#### 2.0.0-24
* Caching HTML response by pathname instead of entire path with query/hash

#### 2.0.0-23
* Remove `cache-control` header from WordPress preview link

#### 2.0.0-22
* Added `CACHE_CONTROL_MAX_AGE` env var

#### 2.0.0-21
* Exposed API response data to `onPageUpdate` function
* Fixed issue when user clicks on multiple links before the previous page had finished loading
* Removed `async-props` from vendor bundle

#### 2.0.0-20
* Added `core-js/modules/es6.symbol` polyfill to main bundle
* Updated Hapi to `~16.5.2`

#### 2.0.0-19
* Internal Hapi server now uses `301` redirects instead of `308` in order to support IE10/11

#### 2.0.0-18
* Pinned Glamor version exactly to prevent regression around `attr()`

#### 2.0.0-17
* Upgraded to webpack `3.4.0`
* Rewrote cache logs, updated some `debug` calls to `silly`
* Fixed homepage caching that was caused by stripping trailing slash from routes
* Implemented `redirectsEndpoint` functionality, add a path to a `redirects.json` and Hapi will fetch the file and set all redirects

#### 2.0.0-16
* Added `forceHttps` option to register `hapi-require-https` plugin
* Updated redirect to retain `request.url.search`
* Added year long cache for static assets served from `_scripts`
* Removed case sensitivity from Hapi routes
* Fixed static routes (without an endpoint) returning a `404` when they should return `200`

#### 2.0.0-15
* Stripping trailing slash from routes

#### 2.0.0-14
* Removed `babel` utility build step, all files are now `node 6` compatible
* Implemented `http-status` lib for correct status code/messages
* Improved log consistency
* Added ability to read redirects from a `redirects.json` file in the root of the project
* Added default security option to Hapi https://hapijs.com/api#route-configuration
* Returning response data to `customError` component when available
* Normalised API response and error handling between client and server
* Updated error view to show `Missing` only when the component is missing, if `CustomError` is declared then that will always show
* Removed `tapestry-log.log` output from Winston as the implementation was buggy and unfinished
* Improved how `progressIndicator` handles long loading times, increased height to `3px`


src/server.js
import fs from 'fs-extra'
import path from 'path'
import { match } from 'react-router'

import { Server } from 'hapi'
import h2o2 from 'h2o2'
import Inert from 'inert'
import { loadPropsOnServer } from 'async-props'
import { has } from 'lodash'

import DefaultRoutes from './default-routes'

import { renderHtml } from './render'
import { success, error } from './logger'


export default class Tapestry {

  constructor ({ config, cwd, env, configPath }, { silent } = {}) {
    // allow access from class
    this.config = config.default
    this.configPath = configPath
    this.context = cwd
    this.env = env
    this.silent = silent
    // override defaults
    this.assets = null
    if (env === 'production')
      this.assets = fs.readJsonSync(path.resolve(cwd, '.tapestry/assets.json'))
    // run server
    this.bootServer()
    this.registerProxies()
    // set routes
    this.routeApi()
    this.routeStatic()
    this.routeDynamic()
    this.startServer()
  }

  registerProxies () {
    if (this.config.proxyPaths)
      this.config.proxyPaths.map(this.routeProxy.bind(this))
  }

  bootServer () {
    // create new Hapi server and register required plugins
    this.server = new Server()
    this.server.register([h2o2, Inert])
    this.server.connection({
      host: this.config.host || '0.0.0.0',
      port: this.config.port || process.env.PORT || 3030
    })
    this.config.serverUri = this.server.info.uri
  }
  startServer () {
    // run server
    this.server.start(err => {
      if (err) error(err)
      if (!this.silent) success(`Server ready: ${this.server.info.uri}`)
    })
  }

  routeApi () {
    this.server.route({
      method: 'GET',
      path: `/api/v1/{query*}`,
      handler: {
        proxy: {
          mapUri: (request, callback) => {
            const url = this.config.siteUrl + '/wp-json/wp/v2/' + request.params.query + request.url.search
            callback(null, url)
          }
        }
      }
    })
  }
  routeProxy (path) {
    this.server.route({
      method: 'GET',
      path: `${path}`,
      handler: {
        proxy: {
          uri: this.config.siteUrl + path,
          passThrough: true
        }
      }
    })
  }
  routeStatic () {
    this.server.route({
      method: 'GET',
      path: '/_scripts/{param*}',
      handler: {
        directory: {
          path: '_scripts'
        }
      }
    })
    this.server.route({
      method: 'GET',
      path: '/public/{param*}',
      handler: {
        directory: {
          path: 'public'
        }
      }
    })
  }
  routeDynamic () {
    this.server.route({
      method: 'GET',
      path: '/{path*}',
      handler: (request, reply) => {

        if (this.env === 'development') {
          // clear the cached modules in memory
          Object
            .keys(require.cache)
            .forEach(key => delete require.cache[key])
          // re-require that config from the parent project to get the latest
          const updatedConfig = require(this.configPath).default
          // combine with previously set values (serverUri etc.)
          Object.assign(this.config, updatedConfig)
        }

        match({
          routes: DefaultRoutes(this.config.components || {}),
          location: request.url.path
        }, (err, redirectLocation, renderProps) => {

          // define global deets for nested components
          const loadContext = this.config

          // 404 if non-matched route
          if (!renderProps) {
            return reply(
              renderHtml({
                loadContext,
                assets: this.env === 'production' ? this.assets : null
              })
            ).code(404)
          }

          // 500 if error from Router
          if (err) {
            error(err)
            return reply(err.message).code(500)
          }

          // 301/2 if redirect
          if (redirectLocation)
            return reply.redirect(redirectLocation)

          // get all the props yo
          loadPropsOnServer(renderProps, loadContext, (err, asyncProps) => {

            let status = 200

            const failApi = has(asyncProps.propsArray[0], 'data.data.status')
            const failRoute = renderProps.routes[1].path === '*'

            if (failApi || failRoute)
              status = 404

            // 500 if error from AsyncProps
            if (err) {
              error(err)
              return reply(err).code(500)
            }

            // 200 with rendered HTML
            reply(
              renderHtml({
                renderProps,
                loadContext,
                asyncProps,
                assets: this.env === 'production' ? this.assets : null
              })
            ).code(status)
          })
        })
      }
    })
  }
}
